File: PHP/Algorithm/34_在排序数组中查找元素的第一个和最后一个位置.php
<?php
/**
 * 题目
 *     给定一个按照升序排列的整数数组 nums，和一个目标值 target。找出给定目标值在数组中的开始位置和结束位置。
 *     你的算法时间复杂度必须是 O(log n) 级别。
 *     如果数组中不存在目标值，返回 [-1, -1]。
 * 示例 1:
 *     输入: nums = [5,7,7,8,8,10], target = 8
 *     输出: [3,4]
 * 示例 2:
 *     输入: nums = [5,7,7,8,8,10], target = 6
 *     输出: [-1,-1]
 */
class Solution
{
    /**
     * 思路
     *     二分查找，在找到位置后，进行前后的数据校验比对
     * 耗时
     *     执行用时 :124 ms, 在所有 PHP 提交中击败了7.41%的用户
     *     内存消耗 :19.1 MB, 在所有 PHP 提交中击败了6.25%的用户
     */
    function searchRange($nums, $target)
    {
        if (!$nums) {
            return [-1, -1];
        }

        // 开始二分
        $hei    = count($nums);
        $min    = 0;
        $outPut = [];

        while ($hei > $min) {

            $middle = $min + floor(($hei - $min) / 2);

            if ($nums[$middle] == $target) {
                $outPut[] = $middle;

                $next = $middle + 1;
                $prev = $middle - 1;

                while (isset($nums[$next]) && $nums[$next] === $target) {
                    $outPut[] = $next;
                    $next++;
                }

                while (isset($nums[$prev]) && $nums[$prev] === $target) {
                    array_unshift($outPut, $prev);
                    $prev--;
                }

                break;
            }
            elseif ($nums[$middle] > $target) {
                $hei = $middle;
            }
            elseif ($nums[$middle] < $target) {
                $min = $middle + 1;
            }
        }

        $outPut = [
            $outPut ? $outPut[0] : -1,
            $outPut ? end($outPut) : -1,
        ];

        return $outPut;
    }
}

$s = new Solution();
var_dump($s->searchRange([2,2], 2));

File: Go/Algorithm/34_在排序数组中查找元素的第一个和最后一个位置.go
package main

import (
	"fmt"
)

func main() {

	fmt.Println(searchRange([]int{0,0,2,3,4,4,4,5}, 5))

}

//思路
//	先二分法找到节点位置(在哪个位置找到都行),在中心扩散找到边界
//	最好log2n最差n
//耗时
//	执行用时： 8 ms , 在所有 Go 提交中击败了 88.67% 的用户
//  内存消耗： 3.9 MB , 在所有 Go 提交中击败了 60.62% 的用户
func searchRange(nums []int, target int) []int {

	left, right := -1, -1

	low, hei := 0, len(nums) - 1

	for low <= hei {
		middle := (hei - low) / 2 + low

		if nums[middle] > target {
			hei--
		} else if nums[middle] < target {
			low = low + 1
		} else {

			left, right = middle, middle

			//确定左边界
			left_tag := 1
			for {
				if middle == 0 {
					left = middle
					break
				}

				if middle - left_tag < 0 {
					break
				}

				if nums[middle-left_tag] != nums[middle] {
					break
				}

				left = middle - left_tag
				left_tag++
			}

			//确定右边界
			right_tag := 1
			for {
				if middle == len(nums) - 1 {
					right = middle
					break
				}

				if middle + right_tag > len(nums) - 1 {
					break
				}

				if nums[middle+right_tag] != nums[middle] {
					break
				}

				right = middle + right_tag
				right_tag++
			}

			break
		}
	}


	return []int{left, right}
}